### PR TITLE
Add type checking to Detect

### PR DESCRIPTION
**Pull request by JakeMathews**
A lot of simple error's that would be caught at compile time are slipping through the cracks and becoming runtime errors. By adding the @TypeChecked annotation to the top of every class, we can help prevent a lot of these mistakes. 

The pros:
A lot of the runtime errors we experience will now surface at runtime
The code will be more defined and easier to read

The cons:
Add @TypeChecked annotation to every class
Some classes may need to be rewritten or tweaked to comply

There is another annotation called @CompileStatic which in addition to type checking, it will statically compile the class. I have opted not to do this because it prevents us from accessing private or read-only fields. @CompileStatic also prevents us from using the ? operator as a null check in some cases, causing null pointer exceptions at runtime where ever they are used.

I have caught a few errors that would have caused run time errors which are fixed in this branch.
I think overall this is a net benefit to the quality and stability of the code, even if it is slightly less flexible

### CODE REVIEW

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/PearBomTool.groovy [hunk 1]
```diff
@@ -74,9 +77,9 @@ class PearBomTool extends BomTool {
         ExecutableOutput pearDependencies = executableRunner.runExe(pearExePath, 'package-dependencies', PACKAGE_XML_FILENAME)
 
         File packageFile = detectFileManager.findFile(sourcePath, PACKAGE_XML_FILENAME)
-        def packageXml = new XmlSlurper().parseText(packageFile.text)
+        GPathResult packageXml = new XmlSlurper().parseText(packageFile.text)
         String rootName = packageXml.name
-        String rootVersion = packageXml.version.release
+        String rootVersion = packageXml.version.result
```
Comment (jrichardBD) [inline src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/PearBomTool.groovy]: Why did you change this from `.release` to `.result`? If you look at the package.xml file in src/test/resources for Pear, it should be `.release`

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/npm/NpmCliDependencyFinder.groovy [hunk 2]
```diff
@@ -35,10 +37,12 @@ import com.blackducksoftware.integration.hub.detect.nameversion.NameVersionNodeT
 import com.blackducksoftware.integration.hub.detect.util.DetectFileManager
 import com.blackducksoftware.integration.hub.detect.util.executable.ExecutableRunner
 import com.google.gson.Gson
+import com.google.gson.JsonElement
 import com.google.gson.JsonObject
 import com.google.gson.JsonParser
 
 @Component
+
```
Comment (jrichardBD) [inline src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/npm/NpmCliDependencyFinder.groovy]: delete this line

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/util/FileFinder.groovy [hunk 3]
```diff
@@ -118,36 +118,36 @@ class FileFinder {
         return files
     }
 
-    private File[] findDirectoriesContainingDirectoriesToDepth(final File sourceDirectory, final String directoryPattern, int maxDepth) {
-        findDirectoriesContainingDirectoriesToDepthRecursive(sourceDirectory, directoryPattern, 0, maxDepth);
+    public File[] findDirectoriesContainingDirectoriesToDepth(final File sourceDirectory, final String directoryPattern, int maxDepth) {
```
Comment (jrichardBD) [inline src/main/groovy/com/blackducksoftware/integration/hub/detect/util/FileFinder.groovy]: Why did you change this to public?

**Comment by gkillough** [comment]
Why is the TypeChecked annotation fully qualified every time as opposed to imported? I understand that an import adds an extra line, but this seems to break a nearly universal convention of Java/Groovy.

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/util/FileFinder.groovy (hunk 3 above)
Comment (JakeMathews) [reply src/main/groovy/com/blackducksoftware/integration/hub/detect/util/FileFinder.groovy]: It was called by detect configuration. With the @CompileStatic annotation you cannot access private methods. Switching to @TypeChecked means I can revert this back to a private method

**Comment by JakeMathews** [comment]
When pasting it, sometimes it would screw up the imports and it became very frustrating having to fix the imports in every file

**Comment by JakeMathews** [comment]
@jrichardBD All set?
